### ISSUE TITLE
DeepChem Model Saving/Reloading Triage

### ISSUE THREAD

**Issue by rbharath** [deepchem/deepchem#2217]
DeepChem has a large collection of models not all of which can be saved/reloaded. #2151 adds some first correctness tests for some of our models, but there are many more we need to fix. Let's use this thread to track the needed work

- [ ] `dc.models.AtomicConvModel`
- [ ] `dc.models.ChemCeption`
- [X] `dc.models.CNN`
- [ ] `dc.models.DTNNModel`
- [ ] `dc.models.DAGModel`
- [ ] `dc.models.GraphConvModel`
- [ ] `dc.models.MPNNModel`
- [X] `dc.models.MultitaskClassifier`
- [X] `dc.models.MultitaskRegressor`
- [ ] `dc.model.MultitaskFitTransformRegressor`
- [ ] `dc.models.MultitaskIRVClassifier`
- [X] `dc.models.ProgressiveMultitaskClassifier`
- [X] `dc.models.ProgressiveMultitaskRegressor`
- [X] `dc.models.RobustMultitaskClassifier`
- [X] `dc.models.RobustMultitaskRegressor`
- [ ] `dc.models.ScScoreModel`
- [ ] `dc.models.SeqToSeq`
- [ ] `dc.models.Smiles2Vec`
- [ ] `dc.models.TextCNNModel`
- [ ] `dc.models.CGCNNModel`
- [ ] `dc.models.GATModel`
- [X] `dc.models.SklearnModel`
- [ ] `dc.models.XgboostModel`

Some of these models may already work and we just need to test them. Others have subtle saving/reloading issues. It's a priority for me to get this figured out since we can't use DeepChem models in any critical settings without saving/loading guarantees. For any folks interested in helping, feel free to grab a model and try to save/reload it!

**Comment by nesanders** [comment]
Hi @rbharath , I've been investigating issues with saving/reloading of `GraphConvModel directly from checkpoints`.  My observation has been that the model restored from disk will *not* make predictions consistent with the original model kept in memory, and also that the reloaded weighs are usually identical between the original and restored models *except* for the first and final kernel variables of the `GraphConv` layers.  Sometimes the first bias variable seems to vary also.

Here's a MWE with versions specified and some example output included: https://gist.github.com/nesanders/316dcde02e1f07c82e5a46805e0f0fc2

**Comment by rbharath** [comment]
@nesanders Thanks for the report! I've been struggling with getting this one working too. I tried applying the fix @hsjang001205 used for Weave/DAG but that didn't appear to do the trick. As a hack workaround, reloading it twice seems to do the trick, but the first reload fails and I'm not sure why.

@hsjang001205 If you're interested, please take a crack at it too! I'll keep looking at it as well and try to figure out what's happening

**Comment by rbharath** [comment]
Perhaps this old TF issue (https://github.com/tensorflow/tensorflow/issues/34298) is involved? It feels like the bug on `GraphConvModel` is perhaps TF Eager specific since `GraphConvModel` is currently the only DeepChem model that uses TF eager mode

**Comment by rbharath** [comment]
Here's a workaround test that passes for `GraphConvModel` where the trick is to `restore()`, then `predict(...)` then `restore()` again. I'm not sure why this works but perhaps some shapes aren't resolved correctly to start?

```
  import deepchem as dc
  featurizer = dc.feat.ConvMolFeaturizer()
  tasks = ["outcome"]
  n_tasks = len(tasks)
  mols = ["C", "CO", "CC"]
  n_samples = len(mols)
  X = featurizer(mols)
  y = np.array([0, 1, 0])
  dataset = dc.data.NumpyDataset(X, y)

  classification_metric = dc.metrics.Metric(
      dc.metrics.roc_auc_score, np.mean, mode="classification")

  batch_size = 10
  model_dir = tempfile.mkdtemp()
  model = dc.models.GraphConvModel(
      len(tasks),
      batch_size=batch_size,
      batch_normalize=False,
      mode='classification',
      model_dir=model_dir)

  model.fit(dataset, nb_epoch=10)
  scores = model.evaluate(dataset, [classification_metric])
  assert scores[classification_metric.name] >= 0.9


  # Reload trained Model
  reloaded_model = dc.models.GraphConvModel(
      len(tasks),
      batch_size=batch_size,
      batch_normalize=False,
      mode='classification',
      model_dir=model_dir)
  reloaded_model.restore()

  # Check predictions match on random sample
  predmols = ["CCCC", "CCCCCO", "CCCCC"]
  Xpred = featurizer(predmols)
  predset = dc.data.NumpyDataset(Xpred)
  origpred = model.predict(predset)
  reloadpred = reloaded_model.predict(predset)
  #assert np.all(origpred == reloadpred)

  # Try re-restore
  reloaded_model.restore()
  reloadpred = reloaded_model.predict(predset)
  assert np.all(origpred == reloadpred)
```

**Comment by hsjang001205** [comment]
@rbharath I found the part where the error occurred. I will try to more debug and PR.

**Comment by ncfrey** [comment]
Put up PR #2258 to add reloading capability and reloading tests for `dc.models.NormalizingFlowModel`!

**Comment by icewaterice** [comment]
I meet some trouble, and one of them that I can't handle it.

It is "**_FailedPreconditionError: Attempting to use uninitialized value weave_layer_8/Variable_8
	 [[{{node _retval_weave_layer_8/Variable_8_0_10}}]]_**"

Is tensorflow's version didn't suit deepchem or other issues?
By the way, the code I was followed yours at this pages, and I do some changes. 

```
`
import deepchem as dc
import numpy as np
from rdkit import Chem
molecules= []
featurizer = dc.feat.WeaveFeaturizer()
tasks = ["outcome"]
mols = ["C", "CO", "CC"]
for i in mols:
    molecules.append(Chem.MolFromSmiles(i))
n_samples = len(mols)
X = featurizer(molecules)
y = np.random.randint(2, size=(n_samples, n_tasks))
dataset = dc.data.NumpyDataset(X, y)

classification_metric = dc.metrics.Metric(dc.metrics.roc_auc_score)

batch_size = 10

model_dir = './model_dir'
model = dc.models.WeaveModel(
      n_tasks,
      batch_size=batch_size,
      learning_rate=0.0003,
      mode="classification",
      dropouts=0.0,
      model_dir=model_dir)


model.fit(dataset, nb_epoch=30)


scores = model.evaluate(dataset, [classification_metric])


model.model.save('./model_save')


`
```

**Comment by rbharath** [comment]
@icewaterice Instead of using `model.model.save()` can you try `model.save()` and `model.restore()`? You're going down to the underlying Keras model save functionality which we haven't tested. If you use `WeaveModel`'s save functionality we should be fine since Travis is passing there.

**Comment by icewaterice** [comment]
@rbharath When I try model.save(), it raised an error which is '**_NotImplementedError:_**'.  I thought somethins that I miss, but I can't find this function in models when I look for it.

**Comment by peastman** [comment]
The `save()` method is mostly obsolete.  Model parameters automatically get saved during fitting, or you can use `save_checkpoint()`.

**Comment by nesanders** [comment]
@rbharath I was curious if you were ultimately able to resolve this issue?  Is it addressed in newer versions of Deepchem?

**Comment by rbharath** [comment]
@nesanders `GraphConvModel` saving/reloading is all fixed in DeepChem 2.4.0. There are still a couple of straggler models (the ones without checks above) that we will fix in a future version!

**Comment by nesanders** [comment]
Thank you @rbharath .  I can confirm in my recent testing that the MWE I posted above passes when run with 2.4.0.  

Note to anyone using the MWE - you must change the `model_instance` attribute to `model` for it to work with the new version.
